In [30]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [31]:
#Load dataset
df = pd.read_excel("Online retail.xlsx" , header=None)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [32]:
df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0
0  shrimp,almonds,avocado,vegetables mix,green gr...
1                             burgers,meatballs,eggs
2                                            chutney
3                                     turkey,avocado
4  mineral water,milk,energy bar,whole wheat rice...

#Pre-processing

In [33]:
df[0].isnull().sum()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0

In [34]:
df.duplicated().sum()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


2325

No null values are available. Since 2325 duplicates are indicated, they are probably not genuine duplicates. Considering it is a retail dataset people may have bought the same items at different days or times. Hence we are not removing them.

In [35]:
#Split the Comma-Separated Strings and convert to Transaction format
transactions = []
for index, row in df.iterrows():
    items_string = row.iloc[0]  # Select the first column (index 0) using iloc
    items_list = [item.strip() for item in items_string.split(',')]  # Remove leading/trailing whitespace
    transactions.append(items_list)

transactions

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
 ['low fat yogurt'],
 ['whole wheat pasta', 'french fries'],
 ['soup', 'light cream', 'shallot'],
 ['frozen vegetables', 'spaghetti', 'green tea'],
 ['french fries'],
 ['eggs', 'pet food'],
 ['cookies'],
 ['turkey', 'burgers', 'mineral water', 'eggs', 'cooking oil'],
 ['spaghetti', 'champagne', 'cookies'],
 ['mineral water', 'salmon'],
 ['mineral water'],
 ['shrimp',
  'chocolate',
  'chicken',
  'honey',
  'oil',
  'cooking oil',
  'low fat yogurt'],
 ['turkey', 'eggs'],
 ['turkey',
  'fresh tuna',
  'tomatoes',
  'spagh

In [36]:
te = TransactionEncoder() #Instantiate TransactionEncoder
te_ary = te.fit(transactions).transform(transactions) # fit and transform the transactions
df_encoded = pd.DataFrame(te_ary, columns=te.columns_) #Create the dataframe
df_encoded

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


almonds  antioxydant juice  asparagus  avocado  babies food  bacon  \
0        True               True      False     True        False  False   
1       False              False      False    False        False  False   
2       False              False      False    False        False  False   
3       False              False      False     True        False  False   
4       False              False      False    False        False  False   
...       ...                ...        ...      ...          ...    ...   
7496    False              False      False    False        False  False   
7497    False              False      False    False        False  False   
7498    False              False      False    False        False  False   
7499    False              False      False    False        False  False   
7500    False              False      False    False        False  False   

      barbecue sauce  black tea  blueberries  body spray  ...  turkey  \
0              False      False        False       False  ...   False   
1              False      False        False       False  ...   False   
2              False      False        False       False  ...   False   
3              False      False        False       False  ...    True   
4              False      False        False       False  ...   False   
...              ...        ...          ...         ...  ...     ...   
7496           False      False        False       False  ...   False   
7497           False      False        False       False  ...   False   
7498           False      False        False       False  ...   False   
7499           False      False        False       False  ...   False   
7500           False      False        False       False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0               True        False       False              True   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
7496           False        False       False             False   
7497           False        False       False             False   
7498           False        False       False             False   
7499           False        False       False             False   
7500           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False   True        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False             False  False        False     False  
4                 False              True  False        False     False  
...                 ...               ...    ...          ...       ...  
7496              False             False  False        False     False  
7497              False             False  False        False     False  
7498              False             False  False        False     False  
7499              False             False  False        False     False  
7500              False             False  False         True     False  

[7501 rows x 119 columns]

In [37]:
#Apriori Algorithm
frequent_itemsets = apriori(df_encoded, min_support=0.03, use_colnames=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [38]:
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                            itemsets
0   0.033329                           (avocado)
1   0.033729                          (brownies)
2   0.087188                           (burgers)
3   0.030129                            (butter)
4   0.081056                              (cake)
5   0.046794                         (champagne)
6   0.059992                           (chicken)
7   0.163845                         (chocolate)
8   0.080389                           (cookies)
9   0.051060                       (cooking oil)
10  0.031862                    (cottage cheese)
11  0.179709                              (eggs)
12  0.079323                          (escalope)
13  0.170911                      (french fries)
14  0.043061                       (fresh bread)
15  0.063325                   (frozen smoothie)
16  0.095321                 (frozen vegetables)
17  0.052393                     (grated cheese)
18  0.132116                         (green tea)
19  0.098254                       (ground beef)
20  0.049460                     (herb & pepper)
21  0.047460                             (honey)
22  0.032396                          (hot dogs)
23  0.076523                    (low fat yogurt)
24  0.129583                              (milk)
25  0.238368                     (mineral water)
26  0.065858                         (olive oil)
27  0.095054                          (pancakes)
28  0.042528                            (salmon)
29  0.071457                            (shrimp)
30  0.050527                              (soup)
31  0.174110                         (spaghetti)
32  0.030396                      (tomato juice)
33  0.068391                          (tomatoes)
34  0.062525                            (turkey)
35  0.058526                  (whole wheat rice)
36  0.033196                   (chocolate, eggs)
37  0.034395           (chocolate, french fries)
38  0.032129                   (chocolate, milk)
39  0.052660          (chocolate, mineral water)
40  0.039195              (chocolate, spaghetti)
41  0.036395                (eggs, french fries)
42  0.030796                        (milk, eggs)
43  0.050927               (mineral water, eggs)
44  0.036528                   (eggs, spaghetti)
45  0.033729       (mineral water, french fries)
46  0.035729  (frozen vegetables, mineral water)
47  0.031063          (green tea, mineral water)
48  0.040928        (mineral water, ground beef)
49  0.039195            (spaghetti, ground beef)
50  0.047994               (mineral water, milk)
51  0.035462                   (milk, spaghetti)
52  0.033729           (mineral water, pancakes)
53  0.059725          (mineral water, spaghetti)

In [39]:
frequent_itemsets_sorted = frequent_itemsets.sort_values(by="support", ascending=False)
frequent_itemsets_sorted

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                            itemsets
25  0.238368                     (mineral water)
11  0.179709                              (eggs)
31  0.174110                         (spaghetti)
13  0.170911                      (french fries)
7   0.163845                         (chocolate)
18  0.132116                         (green tea)
24  0.129583                              (milk)
19  0.098254                       (ground beef)
16  0.095321                 (frozen vegetables)
27  0.095054                          (pancakes)
2   0.087188                           (burgers)
4   0.081056                              (cake)
8   0.080389                           (cookies)
12  0.079323                          (escalope)
23  0.076523                    (low fat yogurt)
29  0.071457                            (shrimp)
33  0.068391                          (tomatoes)
26  0.065858                         (olive oil)
15  0.063325                   (frozen smoothie)
34  0.062525                            (turkey)
6   0.059992                           (chicken)
53  0.059725          (mineral water, spaghetti)
35  0.058526                  (whole wheat rice)
39  0.052660          (chocolate, mineral water)
17  0.052393                     (grated cheese)
9   0.051060                       (cooking oil)
43  0.050927               (mineral water, eggs)
30  0.050527                              (soup)
20  0.049460                     (herb & pepper)
50  0.047994               (mineral water, milk)
21  0.047460                             (honey)
5   0.046794                         (champagne)
14  0.043061                       (fresh bread)
28  0.042528                            (salmon)
48  0.040928        (mineral water, ground beef)
49  0.039195            (spaghetti, ground beef)
40  0.039195              (chocolate, spaghetti)
44  0.036528                   (eggs, spaghetti)
41  0.036395                (eggs, french fries)
46  0.035729  (frozen vegetables, mineral water)
51  0.035462                   (milk, spaghetti)
37  0.034395           (chocolate, french fries)
52  0.033729           (mineral water, pancakes)
1   0.033729                          (brownies)
45  0.033729       (mineral water, french fries)
0   0.033329                           (avocado)
36  0.033196                   (chocolate, eggs)
22  0.032396                          (hot dogs)
38  0.032129                   (chocolate, milk)
10  0.031862                    (cottage cheese)
47  0.031063          (green tea, mineral water)
42  0.030796                        (milk, eggs)
32  0.030396                      (tomato juice)
3   0.030129                            (butter)

We have applied a support of 3%. The support values indicate that there is about a 2% chance that the customer will buy mineral water. Items with >1% chance are eggs, spagetti, french fries, chocolate, green tea, and milk.

In [28]:
confidence_rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.04, num_itemsets=len(frequent_itemsets))
confidence_rules_sorted = confidence_rules.sort_values(by="confidence", ascending=False)
confidence_rules_sorted


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents          consequents  antecedent support  \
25        (ground beef)      (mineral water)            0.098254   
27        (ground beef)          (spaghetti)            0.098254   
20  (frozen vegetables)      (mineral water)            0.095321   
29               (milk)      (mineral water)            0.129583   
33           (pancakes)      (mineral water)            0.095054   
35          (spaghetti)      (mineral water)            0.174110   
6           (chocolate)      (mineral water)            0.163845   
15               (eggs)      (mineral water)            0.179709   
30               (milk)          (spaghetti)            0.129583   
34      (mineral water)          (spaghetti)            0.238368   
5                (milk)          (chocolate)            0.129583   
8           (chocolate)          (spaghetti)            0.163845   
12               (milk)               (eggs)            0.129583   
22          (green tea)      (mineral water)            0.132116   
9           (spaghetti)          (chocolate)            0.174110   
26          (spaghetti)        (ground beef)            0.174110   
7       (mineral water)          (chocolate)            0.238368   
14      (mineral water)               (eggs)            0.238368   
11       (french fries)               (eggs)            0.170911   
2           (chocolate)       (french fries)            0.163845   
17          (spaghetti)               (eggs)            0.174110   
31          (spaghetti)               (milk)            0.174110   
16               (eggs)          (spaghetti)            0.179709   
0           (chocolate)               (eggs)            0.163845   
10               (eggs)       (french fries)            0.179709   
28      (mineral water)               (milk)            0.238368   
3        (french fries)          (chocolate)            0.170911   
19       (french fries)      (mineral water)            0.170911   
4           (chocolate)               (milk)            0.163845   
1                (eggs)          (chocolate)            0.179709   
24      (mineral water)        (ground beef)            0.238368   
13               (eggs)               (milk)            0.179709   
21      (mineral water)  (frozen vegetables)            0.238368   
32      (mineral water)           (pancakes)            0.238368   
18      (mineral water)       (french fries)            0.238368   
23      (mineral water)          (green tea)            0.238368   

    consequent support   support  confidence      lift  representativity  \
25            0.238368  0.040928    0.416554  1.747522               1.0   
27            0.174110  0.039195    0.398915  2.291162               1.0   
20            0.238368  0.035729    0.374825  1.572463               1.0   
29            0.238368  0.047994    0.370370  1.553774               1.0   
33            0.238368  0.033729    0.354839  1.488616               1.0   
35            0.238368  0.059725    0.343032  1.439085               1.0   
6             0.238368  0.052660    0.321400  1.348332               1.0   
15            0.238368  0.050927    0.283383  1.188845               1.0   
30            0.174110  0.035462    0.273663  1.571779               1.0   
34            0.174110  0.059725    0.250559  1.439085               1.0   
5             0.163845  0.032129    0.247942  1.513276               1.0   
8             0.174110  0.039195    0.239219  1.373952               1.0   
12            0.179709  0.030796    0.237654  1.322437               1.0   
22            0.238368  0.031063    0.235116  0.986357               1.0   
9             0.163845  0.039195    0.225115  1.373952               1.0   
26            0.098254  0.039195    0.225115  2.291162               1.0   
7             0.163845  0.052660    0.220917  1.348332               1.0   
14            0.179709  0.050927    0.213647  1.188845               1.0   
11            0.179709  0.036395    0.212949  1.184

Weak confidence of 0.04-0.05 indicates that 4% to 5% of the transactions containing the antecedent also contain the consequent. In other words, when the antecedent occurs, the consequent *usually* does *not* occur.

When ground beef is bought, there is a 4% confidence that mineral water will also be bought. Similarly, when gound beef is bought spagetti is also bought.

Chocolate is the antescednt for mineral water, spagetti, french fries, and eggs.

In [29]:
#Generate rules
lift_rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1, num_itemsets=len(frequent_itemsets) )

lift_rules_sorted = lift_rules.sort_values(by="lift", ascending=False)
lift_rules_sorted

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents          consequents  antecedent support  \
22          (spaghetti)        (ground beef)            0.174110   
23        (ground beef)          (spaghetti)            0.098254   
20      (mineral water)        (ground beef)            0.238368   
21        (ground beef)      (mineral water)            0.098254   
19      (mineral water)  (frozen vegetables)            0.238368   
18  (frozen vegetables)      (mineral water)            0.095321   
27          (spaghetti)               (milk)            0.174110   
26               (milk)          (spaghetti)            0.129583   
25               (milk)      (mineral water)            0.129583   
24      (mineral water)               (milk)            0.238368   
4           (chocolate)               (milk)            0.163845   
5                (milk)          (chocolate)            0.129583   
29           (pancakes)      (mineral water)            0.095054   
28      (mineral water)           (pancakes)            0.238368   
30      (mineral water)          (spaghetti)            0.238368   
31          (spaghetti)      (mineral water)            0.174110   
9           (spaghetti)          (chocolate)            0.174110   
8           (chocolate)          (spaghetti)            0.163845   
6           (chocolate)      (mineral water)            0.163845   
7       (mineral water)          (chocolate)            0.238368   
13               (eggs)               (milk)            0.179709   
12               (milk)               (eggs)            0.129583   
3        (french fries)          (chocolate)            0.170911   
2           (chocolate)       (french fries)            0.163845   
15               (eggs)      (mineral water)            0.179709   
14      (mineral water)               (eggs)            0.238368   
11       (french fries)               (eggs)            0.170911   
10               (eggs)       (french fries)            0.179709   
16               (eggs)          (spaghetti)            0.179709   
17          (spaghetti)               (eggs)            0.174110   
1                (eggs)          (chocolate)            0.179709   
0           (chocolate)               (eggs)            0.163845   

    consequent support   support  confidence      lift  representativity  \
22            0.098254  0.039195    0.225115  2.291162               1.0   
23            0.174110  0.039195    0.398915  2.291162               1.0   
20            0.098254  0.040928    0.171700  1.747522               1.0   
21            0.238368  0.040928    0.416554  1.747522               1.0   
19            0.095321  0.035729    0.149888  1.572463               1.0   
18            0.238368  0.035729    0.374825  1.572463               1.0   
27            0.129583  0.035462    0.203675  1.571779               1.0   
26            0.174110  0.035462    0.273663  1.571779               1.0   
25            0.238368  0.047994    0.370370  1.553774               1.0   
24            0.129583  0.047994    0.201342  1.553774               1.0   
4             0.129583  0.032129    0.196094  1.513276               1.0   
5             0.163845  0.032129    0.247942  1.513276               1.0   
29            0.238368  0.033729    0.354839  1.488616               1.0   
28            0.095054  0.033729    0.141499  1.488616               1.0   
30            0.174110  0.059725    0.250559  1.439085               1.0   
31            0.238368  0.059725    0.343032  1.439085               1.0   
9             0.163845  0.039195    0.225115  1.373952               1.0   
8             0.174110  0.039195    0.239219  1.373952               1.0   
6             0.238368  0.052660    0.321400  1.348332               1.0   
7             0.163845  0.052660    0.220917  1.348332               1.0   
13            0.129583  0.030796    0.171365  1.322437               1.0   
12            0.179709  0.030796    0.237654  1.322437               1.0   
3             0.163

Lift values > 2 is indicated by:
spagetti -> ground beef
ground beef -> spagetti


ground beef and mineral water, frozen veggies and mineral water, spagetti and milk, milk and mineral water, chocolate and milk, pancakes and mineral water, spagetti and mineral water, spagetti and chocolate, chocolate and mineral water, eggs and milk, french fries and chocolate are related to each other among other items. These items can be kept in different areas of the store so customers are encouraged to walk around.

Mineral water is present as a consequent in many cases, so it might be a good idea to keep mineral water close to the checkout counter where it is easy to add to the cast. Alternatively, if antecedent is kept near the beginning of the store, mineral water can be kept at the end of the store so that people are more likely to browse through the store while getting the water.

Chocolate can be kept near the frozen section where milk and french fries are kept. The chocolate section can be opposite to the eggs section.

These essentials that people are buying should be kept in the middle or back of the store so that they browse other items while walking there.

Mineral water prices can be increased while providing a sale on all the other items (or vice versa).



>  What is lift and why is it important in association rules?

Lift measures the strength of the association between two items, considering how frequently each item appears in the dataset. Lift helps us understand if the presence of one item (the antecedent) truly increases the likelihood of another item (the consequent) being present, beyond what we'd expect by chance. A lift value greater than 1 suggests that the items are positively associated, meaning they are more likely to be bought together than independently. The higher the lift, the stronger the association. For example, in the code, the lift values that are greater than 2 indicate that customers are buying these items together more often than it would happen by random chance. The lift values are used to make informed business decisions such as store layout or to put items on sale.

> What is support and confidence? How do you calculate them?

Support measures how frequently an item or a set of items (an itemset) appears in the entire dataset. It's essentially the popularity of the itemset. Support is calculated as the number of transactions containing the itemset divided by the total number of transactions.


Confidence measures the likelihood that the consequent item will be purchased given that the antecedent item has already been purchased. In other words, it measures the strength of the implication of one item to another. Confidence is calculated as the support of the itemset (both antecedent and consequent) divided by the support of the antecedent alone.

> What are some limitations or challenges of association rules mining?

Large Datasets: Association rule mining can be computationally expensive when dealing with very large datasets. The algorithm needs to go through many combinations of items.

Many Rules: Applying association rules to even medium-sized datasets will lead to the production of a huge amount of rules. Many rules that are generated may not be actionable or even interesting.

Setting Thresholds: Choosing appropriate values for support and confidence thresholds can be difficult. If the thresholds are too high, many potentially interesting rules might be missed. If the thresholds are too low, a large number of insignificant rules might be generated.

Spurious Associations: Association rules might find relationships that are statistically significant but not practically meaningful. This is why lift is important to add to the analysis. Lift will help eliminate associations that are not as valuable.

Data Quality: The quality of the results depends heavily on the quality of the data. Errors in the data or missing data can lead to incorrect or misleading rules.

Static Nature: Association rule mining typically analyzes static data and might not capture changes in purchasing behavior over time.




